api/src/main/java/com/capitalone/dashboard/auth/token/TokenAuthenticationServiceImpl.java

```diff
@@ -0,0 +1,90 @@
+package com.capitalone.dashboard.auth.token;
+
+import java.util.Collection;
+import java.util.Date;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.commons.lang.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
+import org.springframework.stereotype.Component;
+
+import com.capitalone.dashboard.auth.AuthProperties;
+import com.google.common.collect.Sets;
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.ExpiredJwtException;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+
+@Component
+public class TokenAuthenticationServiceImpl implements TokenAuthenticationService {
+
+	private static final String AUTHORIZATION = "Authorization";
+	private static final String AUTH_PREFIX_W_SPACE = "Bearer ";
+	private static final String AUTH_RESPONSE_HEADER = "X-Authentication-Token";
+	private static final String ROLES_CLAIM = "roles";
+	private static final String DETAILS_CLAIM = "details";
+
+	private AuthProperties tokenAuthProperties;
+	
+	@Autowired
+	public TokenAuthenticationServiceImpl(AuthProperties tokenAuthProperties) {
+		this.tokenAuthProperties = tokenAuthProperties;
+	}
+	
+	@Override
+	public void addAuthentication(HttpServletResponse response, Authentication authentication) {
+		String jwt = Jwts.builder().setSubject(authentication.getName())
+				.claim(DETAILS_CLAIM, authentication.getDetails())
+				.claim(ROLES_CLAIM, getRoles(authentication.getAuthorities()))
+				.setExpiration(new Date(System.currentTimeMillis() + tokenAuthProperties.getExpirationTime()))
+				.signWith(SignatureAlgorithm.HS512, tokenAuthProperties.getSecret()).compact();
```

Comment:
Just so we are on the same page, is your concern having the signature algorithm in the source code, potentially giving some additional information to malicious users?  Sounds like you would rather store that in the database, then pull the algorithm out by id.